src/subject.rs

```diff
@@ -0,0 +1,135 @@
+// Copyright 2022 The NATS Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+trait Subject {
+    fn is_valid(&self) -> bool;
```

Comment:
The point of having a dedicated `Subject` type is that it always represents a valid subject. This way we can rely on the type system instead of depending on regularily checking if a `Subject` is really a valid subject.

Reply:
If we could do it as a true zero cost abstraction agree. Seems like we might be able to..

And again this was me just playing a bit, discovering some more challenging problems with interior pointers etc. lol